tools/topology/topology2/include/components/alh-dai-copier.conf

```diff
@@ -0,0 +1,143 @@
+#
+# ALH DAI gateway copier
+#
+# ALH DAI copier widget. All attributes defined herein are namespaced
+# by alsatplg to "Object.Widget.alh-copier.N.attribute_name"
+#
+# Usage: this component can be used by instantiating it in the parent object. i.e.
+#
+# 	Object.Widget.alh-copier."N" {
+#		dai_index 0
+#		stream_name	"SDW1-Playback"
+#	}
+#
+# Where N is the unique instance number for the alh-copier object within the same alsaconf node.
+
+Class.Widget."alh-copier" {
+	#
+	# Pipeline ID for the alh-copier object
+	#
+	DefineAttribute."index" {}
+
+	# DAI index: this is always 0 when there is no aggregation.
+	DefineAttribute."dai_index" {
+		token_ref	"sof_tkn_dai.word"
+	}
+
+	#
+	# alh-copier object instance
+	#
+	DefineAttribute."instance" {}
+
+	#include common component definition
+	<include/components/widget-common.conf>
+
+	#
+	# alh-copier component UUID
+	#
+	DefineAttribute."uuid" {
+		type "string"
+		# Token set reference name and type
+		token_ref	"sof_tkn_comp.uuid"
+	}
+
+	#
+	# Bespoke Attribute Definitions for alh-copier
+	#
+
+        DefineAttribute."dai_type" {
+                type    "string"
+                token_ref       "sof_tkn_dai.string"
+        }
+
+	DefineAttribute."direction" {
+		type "string"
+		token_ref	"sof_tkn_dai.word"
+		constraints {
+			!valid_values [
+				"playback"
+				"capture"
+			]
+			!tuple_values [
+				0
+				1
+			]
+		}
+	}
+
+	#
+	# cycles per chunk processed
+	#
+	DefineAttribute."cpc" {
+		# Token set reference name and type
+		token_ref	"sof_tkn_comp.word"
+	}
+
+	DefineAttribute."bss_size" {}
+
+	DefineAttribute."is_pages" {
+		# Token set reference name and type
+		token_ref	"sof_tkn_comp.word"
+	}
+
+	DefineAttribute."node_type" {
+		# Token set reference name and type
+		token_ref	"sof_tkn_copier.word"
+		constraints {
+			!valid_values [
+				$ALH_LINK_OUTPUT_CLASS # ALH link output, legacy for SNDW (DSP ->)
+				$ALH_LINK_INPUT_CLASS # ALH link input, legacy for SNDW (DSP <-)
+			]
+		}
+	}
+
+	attributes {
+		#
+		# The alh-copier widget name would be constructed using the stream_name,
+		# direction and dai_index attributes. For ex: "alh-copier.SDW1-Playback.0".
+		#
+		!constructor [
+			"stream_name"
+			"dai_index"
+		]
+
+		#
+		# mandatory attributes that must be provided when the alh-copier class is
+		# instantiated
+		#
+		!mandatory [
+			"no_pm"
+			"num_input_audio_formats"
+			"num_output_audio_formats"
+			"node_type"
+		]
+
+		#
+		# immutable attributes cannot be modified in the object instance
+		# num_input_audio_formats for capture and num_output_formats for playback are
+		# mandatory even though the num_input_pins/num_output_pins are 0 because
+		# these are used to set the DMA buffer format.
+		#
+		!immutable [
+			"uuid"
+			"dai_type"
```

Comment:
isn't the dai_type restricted to ALH only?

Reply:
yes, its set to ALH below and set to immutable here.